<center>
<img src="../../img/ods_stickers.jpg">
## Открытый курс по машинному обучению
Авторы материала: программист-исследователь Mail.ru Group, старший преподаватель Факультета Компьютерных Наук ВШЭ Юрий Кашницкий и Data Scientist в Segmento Екатерина Демидова. Материал распространяется на условиях лицензии [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Можно использовать в любых целях (редактировать, поправлять и брать за основу), кроме коммерческих, но с обязательным упоминанием автора материала.

# <center>Тема 1. Первичный анализ данных с Pandas</center>

**[Pandas](http://pandas.pydata.org)** — это библиотека Python, предоставляющая широкие возможности для анализа данных. С ее помощью очень удобно загружать, обрабатывать и анализировать табличные данные с помощью SQL-подобных запросов. В связке с библиотеками `Matplotlib` и `Seaborn` появляется возможность удобного визуального анализа табличных данных.

In [1]:
import numpy as np
import pandas as pd

Данные, с которыми работают дата саентисты и аналитики, обычно хранятся в виде табличек — например, в форматах `.csv`, `.tsv` или `.xlsx`. Для того, чтобы считать нужные данные из такого файла, отлично подходит библиотека Pandas.

Основными структурами данных в Pandas являются классы `Series` и `DataFrame`. Первый из них представляет собой одномерный индексированный массив данных некоторого фиксированного типа. Второй - это двухмерная структура данных, представляющая собой таблицу, каждый столбец которой содержит данные одного типа. Можно представлять её как словарь объектов типа `Series`. Структура `DataFrame` отлично подходит для представления реальных данных: строки соответствуют признаковым описаниям отдельных объектов, а столбцы соответствуют признакам.

---------

## Демонстрация основных методов Pandas 


### Series

**Создание объекта Series из 5 элементов, индексированных буквами:**

In [3]:
salaries = pd.Series([400, 300, 200, 250], 
              index = ['Andrew', 'Bob', 
                       'Charles', 'Ann']) 
print(salaries)                                                                 

Andrew     400
Bob        300
Charles    200
Ann        250
dtype: int64


In [4]:
salaries[salaries > 250]

Andrew    400
Bob       300
dtype: int64

**Тип данных и переход к numpy**

In [7]:
# type(salaries)
salaries.values

array([400, 300, 200, 250])

**Индексирование возможно в виде s.Name или s['Name'].**

In [8]:
print(salaries.Andrew == salaries['Andrew']) 

True


In [9]:
salaries['Carl'] = np.nan
salaries

Andrew     400.0
Bob        300.0
Charles    200.0
Ann        250.0
Carl         NaN
dtype: float64

In [12]:
salaries.fillna(salaries.median(), inplace=True)

In [13]:
salaries

Andrew     400.0
Bob        300.0
Charles    200.0
Ann        250.0
Carl       275.0
dtype: float64

### Создание и изменение

**Перейдём к рассмотрению объектов типа DataFrame. Такой объект можно создать из массива numpy, указав названия строк и столбцов.**

In [14]:
df1 = pd.DataFrame(np.random.randn(5, 3), 
                   index=['o1', 'o2', 'o3', 'o4', 'o5'], 
                   columns=['f1', 'f2', 'f3'])
df1

f1        f2        f3
o1  0.300996 -1.330479 -0.200509
o2 -1.673926  0.753896 -0.164639
o3  0.493476 -0.245952  1.081855
o4  0.923773 -0.938827  1.674277
o5  0.267231 -1.318264 -0.777485

**Альтернативным способом является создание DataFrame из словаря numpy массивов или списков.**

In [57]:
df2 = pd.DataFrame({'A': np.random.random(5), 
                    'B': ['a', 'b', 'c', 'd', 'e'], 
                    'C': np.arange(5) > 2})
df2

A  B      C
0  0.94  a  False
1  0.11  b  False
2  0.53  c  False
3  0.95  d   True
4  0.69  e   True

**Тип данных и переход к numpy**

In [19]:
type(df2)
type(df2.values)

array([[0.9987286443362157, 'a', False],
       [0.11245381263586307, 'b', False],
       [0.7135968396184204, 'c', False],
       [0.5615694234532927, 'd', True],
       [0.5833069494388132, 'e', True]], dtype=object)

**Обращение к элементам (или целым кускам фрейма):**

In [20]:
df2.at[3, 'B']

'd'

**Изменение элементов и добавление новых:**

In [21]:
df2.at[2, 'B'] = 'f'
df2

A  B      C
0  0.998729  a  False
1  0.112454  b  False
2  0.713597  f  False
3  0.561569  d   True
4  0.583307  e   True

#### Обработка пропущенных значений

In [22]:
df1.at['o2', 'A'] = 2
df1.at['o4', 'C'] = np.nan
df1

f1        f2        f3    A   C
o1  0.300996 -1.330479 -0.200509  NaN NaN
o2 -1.673926  0.753896 -0.164639  2.0 NaN
o3  0.493476 -0.245952  1.081855  NaN NaN
o4  0.923773 -0.938827  1.674277  NaN NaN
o5  0.267231 -1.318264 -0.777485  NaN NaN

**Булева маска для пропущенных значений (True - там, где был пропуск, иначе - False):**

In [23]:
pd.isnull(df1)

f1     f2     f3      A     C
o1  False  False  False   True  True
o2  False  False  False  False  True
o3  False  False  False   True  True
o4  False  False  False   True  True
o5  False  False  False   True  True

**dropna**

In [27]:
df1.dropna(how='all', axis=1)

f1        f2        f3    A
o1  0.300996 -1.330479 -0.200509  NaN
o2 -1.673926  0.753896 -0.164639  2.0
o3  0.493476 -0.245952  1.081855  NaN
o4  0.923773 -0.938827  1.674277  NaN
o5  0.267231 -1.318264 -0.777485  NaN

**Пропуски можно заменить каким-то значением.**

In [28]:
df1.fillna(0)

f1        f2        f3    A    C
o1  0.300996 -1.330479 -0.200509  0.0  0.0
o2 -1.673926  0.753896 -0.164639  2.0  0.0
o3  0.493476 -0.245952  1.081855  0.0  0.0
o4  0.923773 -0.938827  1.674277  0.0  0.0
o5  0.267231 -1.318264 -0.777485  0.0  0.0

### Чтение из файла и первичный анализ

Прочитаем данные и посмотрим на первые 5 строк с помощью метода `head`:

In [29]:
df = pd.read_csv('telecom_churn.csv')

In [32]:
df.head()

State  Account length  Area code International plan Voice mail plan  \
0    KS             128        415                 No             Yes   
1    OH             107        415                 No             Yes   
2    NJ             137        415                 No              No   
3    OH              84        408                Yes              No   
4    OK              75        415                Yes              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1  False  
1                       1  False  
2                       0  False  
3                       2  False  
4                       3  False

В Jupyter-ноутбуках датафреймы `Pandas` выводятся в виде вот таких красивых табличек, и `print(df.head())` выглядит хуже.

Кстати, по умолчанию `Pandas` выводит всего 20 столбцов и 60 строк, поэтому если ваш датафрейм больше, воспользуйтесь функцией `set_option`:

In [33]:
pd.set_option('display.max_columns', 100)
pd.set_option('display.max_rows', 100)

А также укажем значение параметра `presicion` равным 2, чтобы отображать два знака после запятой (а не 6, как установлено по умолчанию.

In [34]:
pd.set_option('precision', 2)

**Посмотрим на размер данных, названия признаков и их типы**

In [36]:
df.shape

(3333, 20)

Видим, что в таблице 3333 строки и 20 столбцов. Выведем названия столбцов:

In [41]:
df.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

In [44]:
df.index

RangeIndex(start=0, stop=3333, step=1)

Чтобы посмотреть общую информацию по датафрейму и всем признакам, воспользуемся методом **`info`**:

In [45]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
State                     3333 non-null object
Account length            3333 non-null int64
Area code                 3333 non-null int64
International plan        3333 non-null object
Voice mail plan           3333 non-null object
Number vmail messages     3333 non-null int64
Total day minutes         3333 non-null float64
Total day calls           3333 non-null int64
Total day charge          3333 non-null float64
Total eve minutes         3333 non-null float64
Total eve calls           3333 non-null int64
Total eve charge          3333 non-null float64
Total night minutes       3333 non-null float64
Total night calls         3333 non-null int64
Total night charge        3333 non-null float64
Total intl minutes        3333 non-null float64
Total intl calls          3333 non-null int64
Total intl charge         3333 non-null float64
Customer service calls    3333 non-null int64


`bool`, `int64`, `float64` и `object` — это типы признаков. Видим, что 1 признак — логический (`bool`), 3 признака имеют тип `object` и 16 признаков — числовые.

**Изменить тип колонки** можно с помощью метода `astype`. Применим этот метод к признаку `Churn` и переведём его в `int64`:

In [46]:
df['Churn'] = df['Churn'].astype('int64')

Метод **`describe`** показывает основные статистические характеристики данных по каждому числовому признаку (типы `int64` и `float64`): число непропущенных значений, среднее, стандартное отклонение, диапазон, медиану, 0.25 и 0.75 квартили.

In [47]:
df.describe()

Account length  Area code  Number vmail messages  Total day minutes  \
count         3333.00    3333.00                3333.00            3333.00   
mean           101.06     437.18                   8.10             179.78   
std             39.82      42.37                  13.69              54.47   
min              1.00     408.00                   0.00               0.00   
25%             74.00     408.00                   0.00             143.70   
50%            101.00     415.00                   0.00             179.40   
75%            127.00     510.00                  20.00             216.40   
max            243.00     510.00                  51.00             350.80   

       Total day calls  Total day charge  Total eve minutes  Total eve calls  \
count          3333.00           3333.00            3333.00          3333.00   
mean            100.44             30.56             200.98           100.11   
std              20.07              9.26              50.71            19.92   
min               0.00              0.00               0.00             0.00   
25%              87.00             24.43             166.60            87.00   
50%             101.00             30.50             201.40           100.00   
75%             114.00             36.79             235.30           114.00   
max             165.00             59.64             363.70           170.00   

       Total eve charge  Total night minutes  Total night calls  \
count           3333.00              3333.00            3333.00   
mean              17.08               200.87             100.11   
std                4.31                50.57              19.57   
min                0.00                23.20              33.00   
25%               14.16               167.00              87.00   
50%               17.12               201.20             100.00   
75%               20.00               235.30             113.00   
max               30.91               395.00             175.00   

       Total night charge  Total intl minutes  Total intl calls  \
count             3333.00             3333.00           3333.00   
mean                 9.04               10.24              4.48   
std                  2.28                2.79              2.46   
min                  1.04                0.00              0.00   
25%                  7.52                8.50              3.00   
50%                  9.05               10.30              4.00   
75%                 10.59               12.10              6.00   
max                 17.77               20.00             20.00   

       Total intl charge  Customer service calls    Churn  
count            3333.00                 3333.00  3333.00  
mean                2.76                    1.56     0.14  
std                 0.75                    1.32     0.35  
min                 0.00                    0.00     0.00  
25%                 2.30                    1.00     0.00  
50%                 2.78                    1.00     0.00  
75%                 3.27                    2.00     0.00  
max                 5.40                    9.00     1.00

Чтобы посмотреть статистику по нечисловым признакам, нужно явно указать интересующие нас типы в параметре `include`. Можно также задать `include`='all', чтоб вывести статистику по всем имеющимся признакам.

In [48]:
df.describe(include=['object', 'bool'])

State International plan Voice mail plan
count   3333               3333            3333
unique    51                  2               2
top       WV                 No              No
freq     106               3010            2411

Для категориальных (тип `object`) и булевых (тип `bool`) признаков  можно воспользоваться методом **`value_counts`**. Посмотрим на распределение нашей целевой переменной — `Churn`:

In [55]:
df['Total day minutes'].value_counts(dropna=False)

174.5    8
159.5    8
154.0    8
175.4    7
162.3    7
183.4    7
198.4    6
185.0    6
153.5    6
155.2    6
184.5    6
215.6    6
168.6    6
194.8    6
170.5    6
146.3    6
167.8    6
178.7    6
157.1    6
142.3    6
189.3    6
206.9    6
206.2    6
203.4    6
189.8    6
216.0    6
227.4    6
197.0    6
181.5    6
150.0    6
175.8    5
148.5    5
189.5    5
141.3    5
209.9    5
145.6    5
183.6    5
221.1    5
154.6    5
174.0    5
193.3    5
160.1    5
146.4    5
182.1    5
177.1    5
194.4    5
175.7    5
191.4    5
179.2    5
239.8    5
        ..
75.0     1
272.4    1
69.4     1
241.8    1
76.1     1
116.1    1
30.9     1
288.5    1
74.5     1
107.0    1
102.7    1
111.0    1
240.9    1
98.8     1
164.8    1
61.2     1
101.0    1
192.2    1
210.0    1
61.9     1
258.0    1
130.0    1
80.2     1
202.2    1
212.4    1
118.6    1
120.7    1
84.7     1
180.2    1
270.3    1
127.4    1
232.9    1
102.1    1
228.4    1
145.9    1
114.1    1
266.9    1
162.4    1
88.1     1
271.7    1

2850 пользователей из 3333 — лояльные, значение переменной `Churn` у них — `0`.

Посмотрим на распределение пользователей по переменной `Area code`. Укажем значение параметра `normalize=True`, чтобы посмотреть не абсолютные частоты, а относительные.

In [52]:
df['Area code'].value_counts(normalize=False)
df['Area code'].value_counts(normalize=True)

415    0.50
510    0.25
408    0.25
Name: Area code, dtype: float64

### Сортировка

`DataFrame` можно отсортировать по значению какого-нибудь из признаков. В нашем случае, например, по `Total day charge` (`ascending=False` для сортировки по убыванию):

In [53]:
df.sort_values(by='Total day charge', 
               ascending=False).head()

State  Account length  Area code International plan Voice mail plan  \
365     CO             154        415                 No              No   
985     NY              64        415                Yes              No   
2594    OH             115        510                Yes              No   
156     OH              83        415                 No              No   
605     MO             112        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
365                       0              350.8               75   
985                       0              346.8               55   
2594                      0              345.3               81   
156                       0              337.4              120   
605                       0              335.5               77   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
365              59.64              216.5               94             18.40   
985              58.96              249.5               79             21.21   
2594             58.70              203.4              106             17.29   
156              57.36              227.4              116             19.33   
605              57.04              212.5              109             18.06   

      Total night minutes  Total night calls  Total night charge  \
365                 253.9                100               11.43   
985                 275.4                102               12.39   
2594                217.5                107                9.79   
156                 153.9                114                6.93   
605                 265.0                132               11.93   

      Total intl minutes  Total intl calls  Total intl charge  \
365                 10.1                 9               2.73   
985                 13.3                 9               3.59   
2594                11.8                 8               3.19   
156                 15.8                 7               4.27   
605                 12.7                 8               3.43   

      Customer service calls  Churn  
365                        1      1  
985                        1      1  
2594                       1      1  
156                        0      1  
605                        2      1

Сортировать можно и по группе столбцов:

In [61]:
df.sort_values(by=['Churn', 'Total day charge'], 
               ascending=[True, False]).head()

State  Account length  Area code International plan Voice mail plan  \
688     MN              13        510                 No             Yes   
2259    NC             210        415                 No             Yes   
534     LA              67        510                 No              No   
575     SD             114        415                 No             Yes   
2858    AL             141        510                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
688                      21              315.6              105   
2259                     31              313.8               87   
534                       0              310.4               97   
575                      36              309.9               90   
2858                     28              308.0              123   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
688              53.65              208.9               71             17.76   
2259             53.35              147.7              103             12.55   
534              52.77               66.5              123              5.65   
575              52.68              200.3               89             17.03   
2858             52.36              247.8              128             21.06   

      Total night minutes  Total night calls  Total night charge  \
688                 260.1                123               11.70   
2259                192.7                 97                8.67   
534                 246.5                 99               11.09   
575                 183.5                105                8.26   
2858                152.9                103                6.88   

      Total intl minutes  Total intl calls  Total intl charge  \
688                 12.1                 3               3.27   
2259                10.1                 7               2.73   
534                  9.2                10               2.48   
575                 14.2                 2               3.83   
2858                 7.4                 3               2.00   

      Customer service calls  Churn  
688                        3      0  
2259                       3      0  
534                        4      0  
575                        1      0  
2858                       1      0

### Индексация и извлечение данных

`DataFrame` можно индексировать по-разному. В связи с этим рассмотрим различные способы индексации и извлечения нужных нам данных из датафрейма на примере простых вопросов.

Для извлечения отдельного столбца можно использовать конструкцию вида `DataFrame['Name']`. Воспользуемся этим для ответа на вопрос: **какова доля нелояльных пользователей в нашем датафрейме?**

**Разные способы обращения к столбцам**

In [69]:
df.Churn.head() # автодополнение
df['Churn'].head() # лучше при использовании констант с именами столбцов и можно получить датафрейм df[['Churn']]
df[['Churn']]

Churn
0         0
1         0
2         0
3         0
4         0
5         0
6         0
7         0
8         0
9         0
10        1
11        0
12        0
13        0
14        0
15        1
16        0
17        0
18        0
19        0
20        0
21        1
22        0
23        0
24        0
25        0
26        0
27        0
28        0
29        0
30        0
31        0
32        0
33        1
34        0
35        0
36        0
37        0
38        0
39        0
40        0
41        1
42        0
43        0
44        0
45        0
46        0
47        0
48        1
49        0
...     ...
3283      0
3284      0
3285      0
3286      0
3287      1
3288      0
3289      0
3290      0
3291      1
3292      0
3293      0
3294      0
3295      0
3296      0
3297      0
3298      0
3299      0
3300      0
3301      1
3302      0
3303      0
3304      1
3305      0
3306      0
3307      0
3308      0
3309      0
3310      0
3311      0
3312      0
3313      0
3314      0
3315      0
3316      0
3317      0
3318      0
3319      0
3320      1
3321      0
3322      1
3323      1
3324      0
3325      0
3326      0
3327      0
3328      0
3329      0
3330      0
3331      0
3332      0

[3333 rows x 1 columns]

In [70]:
df['Churn'].mean()

0.14491449144914492

14,5% — довольно плохой показатель для компании, с таким процентом оттока можно и разориться.

Очень удобной является логическая индексация `DataFrame` по одному столбцу. Выглядит она следующим образом: `df[P(df['Name'])]`, где `P` - это некоторое логическое условие, проверяемое для каждого элемента столбца `Name`. Итогом такой индексации является `DataFrame`, состоящий только из строк, удовлетворяющих условию `P` по столбцу `Name`. 

Воспользуемся этим для ответа на вопрос: **каковы средние значения числовых признаков среди нелояльных пользователей?**

In [75]:
df

State  Account length  Area code International plan Voice mail plan  \
0       KS             128        415                 No             Yes   
1       OH             107        415                 No             Yes   
2       NJ             137        415                 No              No   
3       OH              84        408                Yes              No   
4       OK              75        415                Yes              No   
5       AL             118        510                Yes              No   
6       MA             121        510                 No             Yes   
7       MO             147        415                Yes              No   
8       LA             117        408                 No              No   
9       WV             141        415                Yes             Yes   
10      IN              65        415                 No              No   
11      RI              74        415                 No              No   
12      IA             168        408                 No              No   
13      MT              95        510                 No              No   
14      IA              62        415                 No              No   
15      NY             161        415                 No              No   
16      ID              85        408                 No             Yes   
17      VT              93        510                 No              No   
18      VA              76        510                 No             Yes   
19      TX              73        415                 No              No   
20      FL             147        415                 No              No   
21      CO              77        408                 No              No   
22      AZ             130        415                 No              No   
23      SC             111        415                 No              No   
24      VA             132        510                 No              No   
25      NE             174        415                 No              No   
26      WY              57        408                 No             Yes   
27      MT              54        408                 No              No   
28      MO              20        415                 No              No   
29      HI              49        510                 No              No   
30      IL             142        415                 No              No   
31      NH              75        510                 No              No   
32      LA             172        408                 No              No   
33      AZ              12        408                 No              No   
34      OK              57        408                 No             Yes   
35      GA              72        415                 No             Yes   
36      AK              36        408                 No             Yes   
37      MA              78        415                 No              No   
38      AK             136        415                Yes             Yes   
39      NJ             149        408                 No              No   
40      GA              98        408                 No              No   
41      MD             135        408                Yes             Yes   
42      AR              34        510                 No              No   
43      ID             160        415                 No              No   
44      WI              64        510                 No              No   
45      OR              59        408                 No             Yes   
46      MI              65        415                 No              No   
47      DE             142        408                 No              No   
48      ID             119        415                 No              No   
49      WY              97        415                 No             Yes   
...    ...             ...        ...                ...             ...   
3283    AK              99        510                

In [78]:
df[(df['Churn'] == 1) & (df['Total day charge'] > 20)]

State  Account length  Area code International plan Voice mail plan  \
10      IN              65        415                 No              No   
15      NY             161        415                 No              No   
33      AZ              12        408                 No              No   
41      MD             135        408                Yes             Yes   
48      ID             119        415                 No              No   
54      WY              87        415                 No              No   
57      CO             121        408                 No             Yes   
69      TX             150        510                 No              No   
76      DC              82        415                 No              No   
84      TX             106        510                 No              No   
86      IN              94        408                 No              No   
89      NJ             128        415                 No              No   
91      LA             155        415                 No              No   
98      MS             120        408                 No              No   
99      ID              77        415                 No              No   
115     ME              36        510                Yes             Yes   
117     NV              74        415                 No              No   
127     MS              61        510                 No             Yes   
144     VT             117        408                Yes              No   
156     OH              83        415                 No              No   
197     TX             208        510                 No              No   
198     ME             131        510                Yes             Yes   
214     FL              70        510                Yes              No   
218     MD             136        415                 No              No   
230     MD              93        408                Yes              No   
235     MN             139        510                 No              No   
241     NV             137        415                Yes              No   
244     VA             108        408                 No              No   
250     OR              33        415                 No             Yes   
277     SD             144        408                 No             Yes   
289     OK              89        510                 No              No   
293     CT              96        415                 No             Yes   
301     FL             127        415                 No              No   
302     RI             121        408                 No              No   
306     VT             113        415                 No              No   
307     IN             122        408                 No              No   
310     MN              62        415                 No              No   
319     SD             128        510                Yes             Yes   
331     MT              54        415                 No              No   
332     AL              86        415                 No              No   
340     GA              89        415                 No              No   
354     MS              64        408                Yes              No   
360     CO             159        408                Yes              No   
365     CO             154        415                 No              No   
372     CA             151        415                Yes              No   
378     MO              95        415                Yes              No   
394     WI              16        510                 No              No   
397     OR              94        415                 No              No   
399     DC              66        415                 No              No   
407     DE             122        510                 No              No   
...    ...             ...        ...                ...             ...   
2961    KY              98        415                

**Важно отличать `and` от `&`, а `or` от `|` и не забывать про скобки в условиях**

Скомбинировав предыдущие два вида индексации, ответим на вопрос: **сколько в среднем в течение дня разговаривают по телефону нелояльные пользователи**?

In [79]:
df[df['Churn'] == 1]['Total day minutes'].mean()

206.91407867494823

**Какова максимальная длина международных звонков среди лояльных пользователей (`Churn == 0`), не пользующихся услугой международного роуминга (`'International plan' == 'No'`)?**

In [83]:
df[(df['Churn'] == 0) & (df['International plan'] == 'No')]['Total intl minutes'].max()

18.9

Датафреймы можно индексировать как по названию столбца или строки, так и по порядковому номеру. Для индексации **по названию** используется метод **`loc`**, **по номеру** — **`iloc`**.

В первом случае мы говорим _«передай нам значения для id строк от 0 до 5 и для столбцов от State до Area code»_, а во втором — _«передай нам значения первых пяти строк в первых трёх столбцах»_. 

В случае `iloc` срез работает как обычно, однако в случае `loc` учитываются и начало, и конец среза.

In [86]:
df1

f1    f2    f3    A   C
o1  0.30 -1.33 -0.20  NaN NaN
o2 -1.67  0.75 -0.16  2.0 NaN
o3  0.49 -0.25  1.08  NaN NaN
o4  0.92 -0.94  1.67  NaN NaN
o5  0.27 -1.32 -0.78  NaN NaN

In [87]:
df1.loc['o3':'o5', 'f2':'A']

f2    f3   A
o3 -0.25  1.08 NaN
o4 -0.94  1.67 NaN
o5 -1.32 -0.78 NaN

In [88]:
df.loc[0:5, 'State':'Area code']
# df.loc['0':'5', 'State':'Area code'] # то же самое

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415
5    AL             118        510

In [89]:
df.iloc[0:5, 0:3]

State  Account length  Area code
0    KS             128        415
1    OH             107        415
2    NJ             137        415
3    OH              84        408
4    OK              75        415

Метод `ix` индексирует и по названию, и по номеру, но он вызывает путаницу, и поэтому был объявлен устаревшим (deprecated).

**Строки и столбцы датафрейма могут быть представлены сериями**

In [94]:
# df.loc[3, :]
# type(df.loc[3, :])
# df.loc[:, 'Area code']
type(df.loc[:, 'Area code'])

pandas.core.series.Series

Если нам нужна первая или последняя строчка датафрейма, пользуемся конструкцией `df[:1]` или `df[-1:]`:

In [95]:
df[-1:]

State  Account length  Area code International plan Voice mail plan  \
3332    TN              74        415                 No             Yes   

      Number vmail messages  Total day minutes  Total day calls  \
3332                     25              234.4              113   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
3332             39.85              265.9               82              22.6   

      Total night minutes  Total night calls  Total night charge  \
3332                241.4                 77               10.86   

      Total intl minutes  Total intl calls  Total intl charge  \
3332                13.7                 4                3.7   

      Customer service calls  Churn  
3332                       0      0

### Применение функций: `apply`, `map` и др.

**Применение функции к каждому столбцу:**

In [98]:
df.apply(np.max)

State                          WY
Account length                243
Area code                     510
International plan            Yes
Voice mail plan               Yes
Number vmail messages          51
Total day minutes         3.5e+02
Total day calls               165
Total day charge               60
Total eve minutes         3.6e+02
Total eve calls               170
Total eve charge               31
Total night minutes         4e+02
Total night calls             175
Total night charge             18
Total intl minutes             20
Total intl calls               20
Total intl charge             5.4
Customer service calls          9
Churn                           1
dtype: object

Метод `apply` можно использовать и для того, чтобы применить функцию к каждой строке. Для этого нужно указать `axis=1`.

**Применение функции к каждой ячейке столбца**

Допустим, по какой-то причине нас интересуют все люди из штатов, названия которых начинаются на 'W'. В данному случае это можно сделать по-разному, но наибольшую свободу дает связка `apply`-`lambda` – применение функции ко всем значениям в столбце.

In [99]:
df[df['State'].apply(lambda state: state[0] == 'W')].head()

State  Account length  Area code International plan Voice mail plan  \
9     WV             141        415                Yes             Yes   
26    WY              57        408                 No             Yes   
44    WI              64        510                 No              No   
49    WY              97        415                 No             Yes   
54    WY              87        415                 No              No   

    Number vmail messages  Total day minutes  Total day calls  \
9                      37              258.6               84   
26                     39              213.0              115   
44                      0              154.0               67   
49                     24              133.2              135   
54                      0              151.0               83   

    Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
9              43.96              222.0              111             18.87   
26             36.21              191.1              112             16.24   
44             26.18              225.8              118             19.19   
49             22.64              217.2               58             18.46   
54             25.67              219.7              116             18.67   

    Total night minutes  Total night calls  Total night charge  \
9                 326.4                 97               14.69   
26                182.7                115                8.22   
44                265.3                 86               11.94   
49                 70.6                 79                3.18   
54                203.9                127                9.18   

    Total intl minutes  Total intl calls  Total intl charge  \
9                 11.2                 5               3.02   
26                 9.5                 3               2.57   
44                 3.5                 3               0.95   
49                11.0                 3               2.97   
54                 9.7                 3               2.62   

    Customer service calls  Churn  
9                        0      0  
26                       0      0  
44                       1      0  
49                       1      0  
54                       5      1

Метод `map` можно использовать и для **замены значений в колонке**, передав ему в качестве аргумента словарь вида `{old_value: new_value}`:

In [100]:
d = {'No' : False, 'Yes' : True}
df['International plan'] = df['International plan'].map(d)
df.head()

State  Account length  Area code  International plan Voice mail plan  \
0    KS             128        415               False             Yes   
1    OH             107        415               False             Yes   
2    NJ             137        415               False              No   
3    OH              84        408                True              No   
4    OK              75        415                True              No   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

Аналогичную операцию можно провернуть с помощью метода `replace`:

In [101]:
df = df.replace({'Voice mail plan': d})
df.head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
1    OH             107        415               False             True   
2    NJ             137        415               False            False   
3    OH              84        408                True            False   
4    OK              75        415                True            False   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  
0                       1      0  
1                       1      0  
2                       0      0  
3                       2      0  
4                       3      0

### Группировка данных

В общем случае группировка данных в Pandas выглядит следующим образом:

```
df.groupby(by=grouping_columns)[columns_to_show].function()
```

1. К датафрейму применяется метод **`groupby`**, который разделяет данные по `grouping_columns` – признаку или набору признаков.
3. Индексируем по нужным нам столбцам (`columns_to_show`). 
2. К полученным группам применяется функция или несколько функций.

**Группирование данных в зависимости от значения признака `Churn` и вывод статистик по трём столбцам в каждой группе.**

In [102]:
columns_to_show = ['Total day minutes', 'Total eve minutes', 'Total night minutes']

df.groupby(['Churn'])[columns_to_show].describe(percentiles=[0.5])

Total day minutes                                   Total eve minutes  \
                  count    mean    std  min    50%    max             count   
Churn                                                                         
0                2850.0  175.18  50.18  0.0  177.2  315.6            2850.0   
1                 483.0  206.91  69.00  0.0  217.6  350.8             483.0   

                                         Total night minutes                 \
         mean    std   min    50%    max               count    mean    std   
Churn                                                                         
0      199.04  50.29   0.0  199.6  361.8              2850.0  200.13  51.11   
1      212.41  51.73  70.9  211.3  363.7               483.0  205.23  47.13   

                            
        min     50%    max  
Churn                       
0      23.2  200.25  395.0  
1      47.4  204.80  354.9

Сделаем то же самое, но немного по-другому, передав в `agg` список функций:

In [110]:
columns_to_show = ['Total day minutes', 'Total eve minutes', 'Total night minutes']

df.groupby(['Churn'])[columns_to_show].agg([np.mean, np.std, np.min, np.max])

Total day minutes                    Total eve minutes               \
                   mean    std amin   amax              mean    std  amin   
Churn                                                                       
0                175.18  50.18  0.0  315.6            199.04  50.29   0.0   
1                206.91  69.00  0.0  350.8            212.41  51.73  70.9   

             Total night minutes                      
        amax                mean    std  amin   amax  
Churn                                                 
0      361.8              200.13  51.11  23.2  395.0  
1      363.7              205.23  47.13  47.4  354.9

Объект groupby

In [116]:
for i, item in df.groupby(['Churn']):
#     print(i)
    item
item

State  Account length  Area code  International plan  Voice mail plan  \
10      IN              65        415               False            False   
15      NY             161        415               False            False   
21      CO              77        408               False            False   
33      AZ              12        408               False            False   
41      MD             135        408                True             True   
48      ID             119        415               False            False   
54      WY              87        415               False            False   
57      CO             121        408               False             True   
69      TX             150        510               False            False   
76      DC              82        415               False            False   
77      NY             144        408               False            False   
84      TX             106        510               False            False   
86      IN              94        408               False            False   
89      NJ             128        415               False            False   
91      LA             155        415               False            False   
98      MS             120        408               False            False   
99      ID              77        415               False            False   
115     ME              36        510                True             True   
117     NV              74        415               False            False   
126     MS             162        415               False            False   
127     MS              61        510               False             True   
144     VT             117        408                True            False   
156     OH              83        415               False            False   
181     AK             126        415               False            False   
197     TX             208        510               False            False   
198     ME             131        510                True             True   
214     FL              70        510                True            False   
218     MD             136        415               False            False   
230     MD              93        408                True            False   
235     MN             139        510               False            False   
241     NV             137        415                True            False   
244     VA             108        408               False            False   
250     OR              33        415               False             True   
258     KS              92        408                True            False   
277     SD             144        408               False             True   
289     OK              89        510               False            False   
293     CT              96        415               False             True   
301     FL             127        415               False            False   
302     RI             121        408               False            False   
306     VT             113        415               False            False   
307     IN             122        408               False            False   
310     MN              62        415               False            False   
319     SD             128        510                True             True   
331     MT              54        415               False            False   
332     AL              86        415               False            False   
340     GA              89        415               False            False   
349     FL             100        510               False            False   
354     MS              64        408                True            False   
360     CO             159        408                True            False   
365     CO             154        415               False            False   
...    ...             ... 

### Сводные таблицы

Допустим, мы хотим посмотреть, как наблюдения в нашей выборке распределены в контексте двух признаков — `Churn` и `Customer service calls`. Для этого мы можем построить **таблицу сопряженности**, воспользовавшись методом **`crosstab`**:

In [111]:
pd.crosstab(df['Churn'], df['International plan'])

International plan  False  True 
Churn                           
0                    2664    186
1                     346    137

In [112]:
pd.crosstab(df['Churn'], df['Voice mail plan'], normalize=True)

Voice mail plan  False  True 
Churn                        
0                 0.60   0.25
1                 0.12   0.02

Мы видим, что большинство пользователей — лояльные и пользуются дополнительными услугами (международного роуминга / голосовой почты).

Продвинутые пользователи `Excel` наверняка вспомнят о такой фиче, как **сводные таблицы** (`pivot tables`). В `Pandas` за сводные таблицы отвечает метод **`pivot_table`**, который принимает в качестве параметров:

* `values` – список переменных, по которым требуется рассчитать нужные статистики,
* `index` – список переменных, по которым нужно сгруппировать данные,
* `aggfunc` — то, что нам, собственно, нужно посчитать по группам — сумму, среднее, максимум, минимум или что-то ещё.

Давайте посмотрим среднее число дневных, вечерних и ночных звонков для разных `Area code`:

In [113]:
df.pivot_table(['Total day calls', 
                'Total eve calls', 
                'Total night calls'], ['Area code'], 
               aggfunc='mean').head(10)

Total day calls  Total eve calls  Total night calls
Area code                                                     
408                 100.50            99.79              99.04
415                 100.58           100.50             100.40
510                 100.10            99.67             100.60

### Преобразование датафреймов

Как и многие другие вещи, добавлять столбцы в `DataFrame` можно несколькими способами.

Например, мы хотим посчитать общее количество звонков для всех пользователей. Создадим объект `total_calls` типа `Series` и вставим его в датафрейм:

In [115]:
total_calls = df['Total day calls'] + df['Total eve calls'] + \
              df['Total night calls'] + df['Total intl calls']
df.insert(loc=len(df.columns), column='Total calls', value=total_calls) 
# loc - номер столбца, после которого нужно вставить данный Series
# мы указали len(df.columns), чтобы вставить его в самом конце
df.head()

ValueError: cannot insert Total calls, already exists

Добавить столбец из имеющихся можно и проще, не создавая промежуточных `Series`:

In [120]:
df['Total charge'] = df['Total day charge'] + df['Total eve charge'] + \
                     df['Total night charge'] + df['Total intl charge']
df.head()

State  Account length  Area code  International plan  Voice mail plan  \
0    KS             128        415               False             True   
1    OH             107        415               False             True   
2    NJ             137        415               False            False   
3    OH              84        408                True            False   
4    OK              75        415                True            False   

   Number vmail messages  Total day minutes  Total day calls  \
0                     25              265.1              110   
1                     26              161.6              123   
2                      0              243.4              114   
3                      0              299.4               71   
4                      0              166.7              113   

   Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
0             45.07              197.4               99             16.78   
1             27.47              195.5              103             16.62   
2             41.38              121.2              110             10.30   
3             50.90               61.9               88              5.26   
4             28.34              148.3              122             12.61   

   Total night minutes  Total night calls  Total night charge  \
0                244.7                 91               11.01   
1                254.4                103               11.45   
2                162.6                104                7.32   
3                196.9                 89                8.86   
4                186.9                121                8.41   

   Total intl minutes  Total intl calls  Total intl charge  \
0                10.0                 3               2.70   
1                13.7                 3               3.70   
2                12.2                 5               3.29   
3                 6.6                 7               1.78   
4                10.1                 3               2.73   

   Customer service calls  Churn  Total calls  Total charge  
0                       1      0          303         75.56  
1                       1      0          332         59.24  
2                       0      0          333         62.29  
3                       2      0          255         66.80  
4                       3      0          359         52.09

Чтобы удалить столбцы или строки, воспользуйтесь методом `drop`, передавая в качестве аргумента нужные индексы и требуемое значение параметра `axis` (`1`, если удаляете столбцы, и ничего или `0`, если удаляете строки):

In [126]:
# избавляемся от созданных только что столбцов
# df = df.drop(['Total charge', 'Total calls'], axis=1) 

# df.drop([1, 2]).head() # а вот так можно удалить строчки
# del df['International plan']
df

State  Account length  Area code  Voice mail plan  Number vmail messages  \
0       KS             128        415             True                     25   
1       OH             107        415             True                     26   
2       NJ             137        415            False                      0   
3       OH              84        408            False                      0   
4       OK              75        415            False                      0   
5       AL             118        510            False                      0   
6       MA             121        510             True                     24   
7       MO             147        415            False                      0   
8       LA             117        408            False                      0   
9       WV             141        415             True                     37   
10      IN              65        415            False                      0   
11      RI              74        415            False                      0   
12      IA             168        408            False                      0   
13      MT              95        510            False                      0   
14      IA              62        415            False                      0   
15      NY             161        415            False                      0   
16      ID              85        408             True                     27   
17      VT              93        510            False                      0   
18      VA              76        510             True                     33   
19      TX              73        415            False                      0   
20      FL             147        415            False                      0   
21      CO              77        408            False                      0   
22      AZ             130        415            False                      0   
23      SC             111        415            False                      0   
24      VA             132        510            False                      0   
25      NE             174        415            False                      0   
26      WY              57        408             True                     39   
27      MT              54        408            False                      0   
28      MO              20        415            False                      0   
29      HI              49        510            False                      0   
30      IL             142        415            False                      0   
31      NH              75        510            False                      0   
32      LA             172        408            False                      0   
33      AZ              12        408            False                      0   
34      OK              57        408             True                     25   
35      GA              72        415             True                     37   
36      AK              36        408             True                     30   
37      MA              78        415            False                      0   
38      AK             136        415             True                     33   
39      NJ             149        408            False                      0   
40      GA              98        408            False                      0   
41      MD             135        408             True                     41   
42      AR              34        510            False                      0   
43      ID             160        415            False                      0   
44      WI              64        510            False                      0   
45      OR              59        408             True                     28   
46      MI              65        415            False                      0   
47      DE             142        408            False                      0   
48      ID             119        41

## Сохранение результата

In [127]:
df.to_csv('df.csv')

In [128]:
df.to_html('df.html')

In [129]:
get_ipython().system('jupyter nbconvert --to script topic1_habr_pandas.ipynb')

[NbConvertApp] Converting notebook topic1_habr_pandas.ipynb to script
[NbConvertApp] Writing 14777 bytes to topic1_habr_pandas.py
